### PR TITLE
Implementing move_to_vector/matrix in State classes to avoid calling qreg() from outside

### PR DESCRIPTION
**Pull request by doichanj**
…x simulators

### Summary
Calling qreg() function from outside will cause problem when we have multiple qregs inside state.
So I added `move_to_vector/matrix` functions in State classes so that State classes with multiple qregs will be able to collect whole vector/matrix in the future release.

### Details and comments
Added `move_to_vector` in `Statevector::State` class
Added `move_to_matrix` in `QubitUnitary::State` , `DensityMatrix::State` and `QubitSuperoperator::State` classes

### CODE REVIEW

File: src/simulators/unitary/unitary_state.hpp [hunk 1]
```diff
@@ -107,6 +107,10 @@ class State : public Base::State<unitary_matrix_t> {
   // Initialize OpenMP settings for the underlying QubitVector class
   void initialize_omp();
 
+  auto move_to_matrix(void)
```
Comment (vvilpas) [inline src/simulators/unitary/unitary_state.hpp:110]: Remove `void`.

File: src/simulators/density_matrix/densitymatrix_state.hpp [hunk 2]
```diff
@@ -129,6 +129,10 @@ class State : public Base::State<densmat_t> {
   // Initialize OpenMP settings for the underlying DensityMatrix class
   void initialize_omp();
 
+  auto move_to_matrix(void)
```
Comment (vvilpas) [inline src/simulators/density_matrix/densitymatrix_state.hpp:132]: Remove `void`.

File: src/simulators/statevector/statevector_state.hpp [hunk 3]
```diff
@@ -142,6 +142,10 @@ class State : public Base::State<statevec_t> {
   // Initialize OpenMP settings for the underlying QubitVector class
   void initialize_omp();
 
+  auto move_to_vector(void)
```
Comment (vvilpas) [inline src/simulators/statevector/statevector_state.hpp:145]: Remove `void`.

File: src/simulators/state.hpp [hunk 4]
```diff
@@ -242,6 +242,12 @@ class State {
   void snapshot_creg_register(const Operations::Op &op, ExperimentResult &result,
                               std::string name = "register") const;
 
+  //add final state to result
+  virtual void add_state_to_data(ExperimentResult &result)
```
Comment (vvilpas) [inline src/simulators/state.hpp:246]: Shouldn't we have overrides of this function in other files such as `state_chunk.hpp`, `statevector_state_chunk.hpp`, etc.

**Review by vvilpas — CHANGES_REQUESTED** [review]

File: src/simulators/superoperator/superoperator_state.hpp [hunk 5]
```diff
@@ -106,6 +106,10 @@ class State : public Base::State<data_t> {
   // Initialize OpenMP settings for the underlying QubitVector class
   void initialize_omp();
 
+  auto move_to_matrix(void)
```
[the remark this answers is not in the document]
Comment (doichanj) [reply src/simulators/superoperator/superoperator_state.hpp:109]: removed

File: src/simulators/density_matrix/densitymatrix_state.hpp (hunk 2 above)
Comment (doichanj) [reply src/simulators/density_matrix/densitymatrix_state.hpp:132]: removed

File: src/simulators/statevector/statevector_state.hpp (hunk 3 above)
Comment (doichanj) [reply src/simulators/statevector/statevector_state.hpp:145]: removed

File: src/simulators/unitary/unitary_state.hpp (hunk 1 above)
Comment (doichanj) [reply src/simulators/unitary/unitary_state.hpp:110]: removed

File: src/simulators/state.hpp (hunk 4 above)
Comment (doichanj) [reply src/simulators/state.hpp:246]: sorry I forgot to remove this, removed

**Review by vvilpas — APPROVED** [review]

**Pull request merged**